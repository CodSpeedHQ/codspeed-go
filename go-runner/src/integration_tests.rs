use itertools::Itertools;
use rstest::rstest;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use tempfile::TempDir;

use crate::results::walltime_results::WalltimeResults;

fn setup_test_project(project_name: &str) -> anyhow::Result<TempDir> {
    let project_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("testdata/projects")
        .join(project_name);
    println!("Project path: {project_path:?}");

    let temp_dir = TempDir::new()?;
    crate::utils::copy_dir_all(&project_path, &temp_dir)?;

    Ok(temp_dir)
}

fn assert_results_snapshots(profile_dir: &Path, project_name: &str) {
    let glob_pattern = profile_dir.join("results");
    if !glob_pattern.exists() {
        eprintln!("No results found for project: {project_name}");
        return;
    }

    let files = std::fs::read_dir(&glob_pattern)
        .expect("Failed to read results directory")
        .filter_map(Result::ok)
        .map(|entry| entry.path())
        .map(|path| {
            let file = std::fs::File::open(&path).unwrap();
            serde_json::from_reader::<_, WalltimeResults>(file).unwrap()
        })
        // Ensure we have the correct order for multiple test executables
        .sorted_by_cached_key(|r| {
            r.benchmarks
                .iter()
                .map(|b| b.metadata.name.clone())
                .sorted()
                .join(";")
        })
        .collect::<Vec<_>>();

    for (i, mut content) in files.into_iter().enumerate() {
        content
            .benchmarks
            .sort_by_cached_key(|b| b.metadata.name.clone());

        let _guard = {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_suffix(format!("{project_name}_{i}"));
            settings.bind_to_scope()
        };

        insta::assert_json_snapshot!(content, {
            ".creator.pid" => "[pid]",
            ".benchmarks[].stats" => "[stats]",
        });
    }
}

#[rstest]
// // #[case::caddy("caddy")]
#[case::fzf("fzf")]
#[case::opentelemetry_go("opentelemetry-go")]
#[case::golang_benchmarks("golang-benchmarks")]
#[case::zerolog("zerolog")]
#[case::zap("zap")]
#[case::hugo("hugo")]
// Currently not producing results:
#[case::fuego("fuego")]
#[case::cli_runtime("cli-runtime")]
fn test_build_and_run(#[case] project_name: &str) {
    let temp_dir = setup_test_project(project_name).unwrap();

    // Mutex to prevent concurrent tests from interfering with CODSPEED_PROFILE_FOLDER env var
    static ENV_MUTEX: Mutex<()> = Mutex::new(());
    let _env_guard = ENV_MUTEX.lock().unwrap_or_else(|e| e.into_inner());

    let profile_dir = temp_dir.path().join("profile");
    unsafe { std::env::set_var("CODSPEED_PROFILE_FOLDER", &profile_dir) };
    if let Err(error) = crate::run_benchmarks(temp_dir.path(), ".") {
        panic!("Benchmarks couldn't run: {error}");
    }

    assert_results_snapshots(&profile_dir, project_name);
}
