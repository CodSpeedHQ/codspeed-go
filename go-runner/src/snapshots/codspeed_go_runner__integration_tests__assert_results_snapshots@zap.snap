---
source: go-runner/src/integration_tests.rs
expression: results
---
[
  {
    "creator": {
      "name": "codspeed-go",
      "version": "[version]",
      "pid": "[pid]"
    },
    "instrument": {
      "type": "walltime"
    },
    "benchmarks": [
      {
        "name": "Benchmark100Fields",
        "uri": "logger_bench_test.go::Benchmark100Fields",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "Benchmark10Fields",
        "uri": "logger_bench_test.go::Benchmark10Fields",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "Benchmark5WithLazysNotUsed",
        "uri": "logger_bench_test.go::Benchmark5WithLazysNotUsed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "Benchmark5WithLazysUsed",
        "uri": "logger_bench_test.go::Benchmark5WithLazysUsed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "Benchmark5WithsNotUsed",
        "uri": "logger_bench_test.go::Benchmark5WithsNotUsed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "Benchmark5WithsUsed",
        "uri": "logger_bench_test.go::Benchmark5WithsUsed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAddCallerAndStacktrace",
        "uri": "logger_bench_test.go::BenchmarkAddCallerAndStacktrace",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAddCallerHook",
        "uri": "logger_bench_test.go::BenchmarkAddCallerHook",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::string::field-only::any",
        "uri": "logger_bench_test.go::BenchmarkAny::string::field-only::any",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::string::field-only::typed",
        "uri": "logger_bench_test.go::BenchmarkAny::string::field-only::typed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::string::log-go::any",
        "uri": "logger_bench_test.go::BenchmarkAny::string::log-go::any",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::string::log-go::typed",
        "uri": "logger_bench_test.go::BenchmarkAny::string::log-go::typed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::string::log::any",
        "uri": "logger_bench_test.go::BenchmarkAny::string::log::any",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::string::log::typed",
        "uri": "logger_bench_test.go::BenchmarkAny::string::log::typed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::stringer::field-only::any",
        "uri": "logger_bench_test.go::BenchmarkAny::stringer::field-only::any",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::stringer::field-only::typed",
        "uri": "logger_bench_test.go::BenchmarkAny::stringer::field-only::typed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::stringer::log-go::any",
        "uri": "logger_bench_test.go::BenchmarkAny::stringer::log-go::any",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::stringer::log-go::typed",
        "uri": "logger_bench_test.go::BenchmarkAny::stringer::log-go::typed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::stringer::log::any",
        "uri": "logger_bench_test.go::BenchmarkAny::stringer::log::any",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkAny::stringer::log::typed",
        "uri": "logger_bench_test.go::BenchmarkAny::stringer::log::typed",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkBoolField",
        "uri": "logger_bench_test.go::BenchmarkBoolField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkByteStringField",
        "uri": "logger_bench_test.go::BenchmarkByteStringField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkDurationField",
        "uri": "logger_bench_test.go::BenchmarkDurationField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkErrorField",
        "uri": "logger_bench_test.go::BenchmarkErrorField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkErrorsField",
        "uri": "logger_bench_test.go::BenchmarkErrorsField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkFloat64Field",
        "uri": "logger_bench_test.go::BenchmarkFloat64Field",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkInt64Field",
        "uri": "logger_bench_test.go::BenchmarkInt64Field",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkIntField",
        "uri": "logger_bench_test.go::BenchmarkIntField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkNoContext",
        "uri": "logger_bench_test.go::BenchmarkNoContext",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkObjectField",
        "uri": "logger_bench_test.go::BenchmarkObjectField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkReflectField",
        "uri": "logger_bench_test.go::BenchmarkReflectField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkStackField",
        "uri": "logger_bench_test.go::BenchmarkStackField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkStringField",
        "uri": "logger_bench_test.go::BenchmarkStringField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkStringerField",
        "uri": "logger_bench_test.go::BenchmarkStringerField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      },
      {
        "name": "BenchmarkTimeField",
        "uri": "logger_bench_test.go::BenchmarkTimeField",
        "config": {
          "warmup_time_ns": null,
          "min_round_time_ns": null,
          "max_time_ns": null,
          "max_rounds": null
        },
        "stats": "[stats]"
      }
    ]
  }
]
